test/remote/gateways/remote_ixopay_test.rb

```diff
@@ -5,36 +5,43 @@ def setup
     @gateway = IxopayGateway.new(fixtures(:ixopay))
 
     @amount = 100
-    @credit_card = credit_card('4000100011112224')
+    @credit_card = credit_card('4111111111111111')
     @declined_card = credit_card('4000300011112220')
+
     @options = {
       billing_address: address,
-      description: 'Store Purchase'
+      shipping_address: address,
+      email: 'test@example.com',
+      description: 'Store Purchase',
+      ip: '192.168.1.1',
     }
   end
 
   def test_successful_purchase
     response = @gateway.purchase(@amount, @credit_card, @options)
+
     assert_success response
-    assert_equal 'REPLACE WITH SUCCESS MESSAGE', response.message
-  end
+    assert_equal 'FINISHED', response.message
+    assert_match(/[0-9a-zA-Z]+(|[0-9a-zA-Z]+)*/, response.authorization)
 
-  def test_successful_purchase_with_more_options
-    options = {
-      order_id: '1',
-      ip: "127.0.0.1",
-      email: "joe@example.com"
-    }
+    assert_equal @credit_card.name,           response.params['card_holder']
+    assert_equal '%02d' % @credit_card.month, response.params['expiry_month']
+    assert_equal @credit_card.year.to_s,      response.params['expiry_year']
+    assert_equal @credit_card.number[0..5],   response.params['first_six_digits']
+    assert_equal 'FINISHED',                  response.params['return_type']
 
-    response = @gateway.purchase(@amount, @credit_card, options)
-    assert_success response
-    assert_equal 'REPLACE WITH SUCCESS MESSAGE', response.message
+    assert_equal @credit_card.number.split(//).last(4).join, response.params['last_four_digits']
```

Comment:
These seem far more in depth than usual for remote tests. What is the advantage of moving these sorts of checks out of unit tests and into remote tests?

Reply:
These assertions are just checking the raw response elements returned by the gateway. So, the tests will fail either if our code starts passing the wrong data, or if the Ixopay response format changes. They weren't really moved out of unit tests (writing the remote tests first is a good strategy that I followed for the new gateway implementation). It would be possible, albeit more complicated, to write unit tests to check the data we're passing to the gateway (perhaps similar to what do for GSFs). However, it wouldn't be possible to check in unit tests that the gateway returns the expected data.

Reply:
I thought that the idea of the test pyramid was that you start writing your tests with unit tests and then work your way out to more complicated things like remote tests.

If the intent of these additional `assert` statements is to ensure that the structure of the gateway response is what we expect, then perhaps it would be worthwhile to extract these out into a new `test_response_format` remote test since these seem to be beyond the scope of what is needed to ensure that a purchase request was successful.